cmd/errtrace/main.go

```diff
@@ -173,27 +183,142 @@ func (cmd *mainCmd) Run(args []string) (exitCode int) {
 		return 1
 	}
 
-	for _, file := range p.Files {
+	files, err := expandPatterns(p.Patterns)
+	if err != nil {
+		cmd.log.Println("errtrace:", err)
+		return 1
+	}
+
+	// Paths will be printed relative to CWD.
+	// Paths outside it will be printed as-is.
+	var workDir string
+	if wd, err := os.Getwd(); err == nil {
+		workDir = wd + string(filepath.Separator)
+	}
+
+	for _, file := range files {
+		display := file
+		if workDir != "" {
+			// Not using filepath.Rel
+			// because we don't want any ".."s in the path.
+			display = strings.TrimPrefix(file, workDir)
+		}
+		if display == "-" {
+			display = "stdin"
+		}
+
 		req := fileRequest{
 			Format:   p.shouldFormat(),
 			Write:    p.Write,
 			List:     p.List,
-			Filename: file,
+			Filename: display,
+			Filepath: file,
 		}
 		if err := cmd.processFile(req); err != nil {
-			cmd.log.Printf("%s:%s", file, err)
+			cmd.log.Printf("%s:%s", display, err)
 			exitCode = 1
 		}
 	}
 
 	return exitCode
 }
 
+// expandPatterns turns the given list of patterns and files
+// into a list of paths to files.
+//
+// Arguments that are already files are returned as-is.
+// Arguments that are patterns are expanded using 'go list'.
+// As a special case for stdin, "-" is returned as-is.
+func expandPatterns(args []string) ([]string, error) {
+	var files, patterns []string
+	for _, arg := range args {
+		if arg == "-" {
+			files = append(files, arg)
+			continue
+		}
+
+		if info, err := os.Stat(arg); err == nil && !info.IsDir() {
+			files = append(files, arg)
+			continue
+		}
+
+		patterns = append(patterns, arg)
+	}
+
+	if len(patterns) > 0 {
+		pkgFiles, err := goListFiles(patterns)
+		if err != nil {
+			return nil, fmt.Errorf("go list: %w", err)
+		}
+
+		files = append(files, pkgFiles...)
+	}
+
+	return files, nil
+}
+
+var _execCommand = exec.Command
+
+func goListFiles(patterns []string) (files []string, err error) {
+	// If we use "go list" to find the files,
+	// it'll only report files that match the default build constraints.
+	//
+	// We want all files so we just request the package directories,
+	// and then walk them ourselves to find all the files.
+	//
+	// The -e flag makes 'go list' include erroneous packages,
+	// e.g. those with all files excluded by build constraints.
```

Comment:
Are you sure about this? With go1.21, I'm not seeing packages that don't match the current build tags being listed,

```
$ go list -e ./...
prashantv/temp/ignore-list
prashantv/temp/ignore-list/linux

$ go list -tags othertag -e ./...
prashantv/temp/ignore-list
prashantv/temp/ignore-list/linux
prashantv/temp/ignore-list/other-tag
```

I think that's fine -- `go fmt` also doesn't seem to handle this case either (I may file an issue for this), but we should update the comment and call that out as a limitation

Reply:
Huh, you're right. `./...` does not match directories with all ignored files, but just plain `.` or `go list` does.
e.g.

```
% go list
package braces.dev/errtrace/cmd/errtrace/testdata/golden: build constraints exclude all Go files in [..]

% go list -e
braces.dev/errtrace/cmd/errtrace/testdata/golden
```